x-pack/metricbeat/module/mssql/performance/performance.go

```diff
@@ -0,0 +1,121 @@
+// you may not use this file except in compliance with the Elastic License.
+
+package performance
+
+import (
+	"database/sql"
+	"fmt"
+	"strings"
+
+	"github.com/elastic/beats/libbeat/common"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/libbeat/common/cfgwarn"
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/metricbeat/mb"
+	"github.com/elastic/beats/x-pack/metricbeat/module/mssql"
+)
+
+type performanceCounter struct {
+	objectName   string
+	instanceName string
+	counterName  string
+	counterValue *int64
+}
+
+// init registers the MetricSet with the central registry as soon as the program
+// starts. The New function will be called later to instantiate an instance of
+// the MetricSet for each host defined in the module's configuration. After the
+// MetricSet has been created then Fetch will begin to be called periodically.
+func init() {
+	mb.Registry.MustAddMetricSet("mssql", "performance", New,
+		mb.DefaultMetricSet(),
+		mb.WithHostParser(mssql.HostParser))
+}
+
+// MetricSet holds any configuration or state information. It must implement
+// the mb.MetricSet interface. And this is best achieved by embedding
+// mb.BaseMetricSet because it implements all of the required mb.MetricSet
+// interface methods except for Fetch.
+type MetricSet struct {
+	mb.BaseMetricSet
+	log     *logp.Logger
+	fetcher *mssql.Fetcher
+	db      *sql.DB
+}
+
+// New creates a new instance of the MetricSet. New is responsible for unpacking
+// any MetricSet specific configuration options if there are any.
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	cfgwarn.Beta("The mssql performance metricset is beta.")
+
+	logger := logp.NewLogger("mssql.performance").With("host", base.HostData().SanitizedURI)
+
+	db, err := mssql.NewConnection(base.HostData().URI)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not create connection to db")
+	}
+
+	return &MetricSet{
+		BaseMetricSet: base,
+		log:           logger,
+		db:            db,
+	}, nil
+}
+
+// Fetch methods implements the data gathering and data conversion to the right format
+// It returns the event which is then forward to the output. In case of an error, a
+// descriptive error must be returned.
+func (m *MetricSet) Fetch(reporter mb.ReporterV2) {
+	var err error
+	var rows *sql.Rows
+	rows, err = m.db.Query("SELECT object_name, counter_name, instance_name, cntr_value FROM sys.dm_os_performance_counters WHERE counter_name = 'SQL Compilations/sec' OR counter_name = 'SQL Re-Compilations/sec' OR counter_name = 'User Connections' OR counter_name = 'Page splits/sec' OR (counter_name = 'Lock Waits/sec' AND instance_name = '_Total') OR counter_name = 'Page splits/sec' OR (object_name = 'SQLServer:Buffer Manager' AND counter_name = 'Page life expectancy') OR counter_name = 'Batch Requests/sec' OR (counter_name = 'Buffer cache hit ratio' AND object_name = 'SQLServer:Buffer Manager') OR (counter_name = 'Lock Waits/sec' and instance_name = '_Total')")
```

Comment:
Works for now. I remember for all these fields 1 column was empty (don't remember which one). Perhaps to simplify the query we could just check if this column is empty?

Reply:
If the query without `WHERE` returns 2000 rows, 1000 are with the empty column. I guess that what you want is to simplify (automatize) the process of adding more fields. What we could do is to ask for everything (those 1000 rows), and take only the values we want from an array but we will still have to _download_ 1000 rows from the db

Reply:
Yes. I would probably download the 1000 rows but I don't know the size of it. Curious what would be less load on the DB.

Reply:
Actually, I have just checked and they are 273 rows and doing some ugly maths by checking the size of each cell, it will be a total of 109200 bytes to download.

Reply:
100KB sounds doable but let's keep it as is for now and "improve" it in case it ever becomes an issue.